Sources/AsyncAlgorithms/Debounce/DebounceStateMachine.swift

```diff
@@ -0,0 +1,757 @@
+            downstreamContinuation: UnsafeContinuation<Element?, Error>,
+            currentElement: (element: Element, deadline: C.Instant)
+        )
+
+        /// The state once any of the upstream sequences threw an `Error`.
+        case upstreamFailure(
+            error: Error
+        )
+
+        /// The state once all upstream sequences finished or the downstream consumer stopped, i.e. by dropping all references
+        /// or by getting their `Task` cancelled.
+        case finished
+    }
+
+    /// The state machine's current state.
+    private var state: State
+    /// The interval to debounce.
+    private let interval: C.Instant.Duration
+    /// The clock.
+    private let clock: C
+
+    init(base: Base, clock: C, interval: C.Instant.Duration) {
+        self.state = .initial(base: base)
+        self.clock = clock
+        self.interval = interval
+    }
+
+    mutating func sequenceDeinitialized() {
+        switch self.state {
+        case .initial:
+            // The references to the sequence were dropped before any iterator was ever created
+            self.state = .finished
+
+        case .waitingForDemand, .demandSignalled, .debouncing, .upstreamFailure:
+            // An iterator was created and we deinited the sequence.
+            // This is an expected pattern and we just continue on normal.
+            // Importantly since we are a unicast sequence no more iterators can be created
+            return
+
+        case .finished:
+            // We are already finished so there is nothing left to clean up.
+            // This is just the references dropping afterwards.
+            return
+        }
+    }
+
+    /// Actions returned by `iteratorInitialized()`.
+    enum IteratorInitializedAction {
+        /// Indicates that a new `Task` should be created that consumes the sequence.
+        case startTask(Base)
+    }
+
+    mutating func iteratorInitialized() -> IteratorInitializedAction {
+        switch self.state {
+        case .initial(let base):
+            // This is the first iterator being created and we need to create our `Task`
+            // that is consuming the upstream sequences.
+            return .startTask(base)
+
+        case .waitingForDemand, .demandSignalled, .debouncing, .upstreamFailure, .finished:
+            fatalError("debounce allows only a single AsyncIterator to be created")
+        }
+    }
+
+    /// Actions returned by `iteratorDeinitialized()`.
+    enum IteratorDeinitializedAction {
+        /// Indicates that the `Task` needs to be cancelled and
+        /// the upstream and clock continuation need to be resumed with a `CancellationError`.
+        case cancelTaskAndUpstreamAndClockContinuations(
+            task: Task<Void, Never>,
+            upstreamContinuation: UnsafeContinuation<Void, Error>?,
+            clockContinuation: UnsafeContinuation<C.Instant, Error>?
+        )
+        /// Indicates that nothing should be done.
+        case none
```

Comment:
almost seems like it is an optional struct and not an enum ;)

Not a big deal but is sus.